### PR TITLE
[LF] improve AstRewriter

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
Allows easy renaming of `packageId` in a Ast  

CHANGELOG_BEGIN
CHANGELOG_END

### CODE REVIEW

File: daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/archive/testing/EncodeV1.scala [hunk 1]
```diff
@@ -528,7 +528,7 @@ private[daml] class EncodeV1(minor: LV.Minor) {
     private def encodeExprBuilder(
         expr0: Expr,
         builder: PLF.Expr.Builder = PLF.Expr.newBuilder(),
-    ): PLF.Expr.Builder = {
+    ): builder.type = {
```
Comment (remyhaemmerle-da) [inline daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/archive/testing/EncodeV1.scala:531]: This ensures the function actually returns `builder` and does not return a new one.

File: daml-lf/encoder/src/test/scala/com/digitalasset/daml/lf/archive/testing/EncodeV1Spec.scala [hunk 2]
```diff
@@ -259,20 +261,15 @@ object EncodeV1Spec {
...
-    val identifierRule: PartialFunction[Identifier, Identifier] = {
-      case Identifier(`hashCode`, name) => Identifier(selfPackageId, name)
+    val replacePkId: PartialFunction[PackageId, PackageId] = { case `hashCode` =>
+      selfPackageId
     }
-    lazy val exprRule: PartialFunction[Expr, Expr] = {
-      case EAbs(binder, body, Some(_)) =>
-        EAbs(normalizer.apply(binder), normalizer.apply(body), None)
-      case ELocation(loc, expr) if loc.packageId == hashCode =>
-        ELocation(loc = loc.copy(packageId = pkgId), normalizer.apply(expr))
```
Comment (remyhaemmerle-da) [inline daml-lf/encoder/src/test/scala/com/digitalasset/daml/lf/archive/testing/EncodeV1Spec.scala:273]: It should be `loc.copy(packageId = selfPackageId)`

File: daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/archive/testing/EncodeV1.scala (hunk 1 above)
Comment (nickchapman-da) [reply daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/archive/testing/EncodeV1.scala:531]: I like this!

**Review by nickchapman-da — COMMENTED** [review]
LGTM

**Review by nickchapman-da — APPROVED** [review]
LGTM

**Pull request merged**